python/lsst/ip/diffim/dipoleFitTask.py

```
@@ -0,0 +1,921 @@
+    # using for algorithm development.
+    _private_version_ = '0.0.4'
+
+    # Create a namedtuple to hold all of the relevant output from the lmfit results
+    resultsOutput = namedtuple('resultsOutput',
+                               ['psfFitPosCentroidX', 'psfFitPosCentroidY',
+                                'psfFitNegCentroidX', 'psfFitNegCentroidY', 'psfFitPosFlux',
+                                'psfFitNegFlux', 'psfFitPosFluxSigma', 'psfFitNegFluxSigma',
+                                'psfFitCentroidX', 'psfFitCentroidY', 'psfFitOrientation',
+                                'psfFitSignaltoNoise', 'psfFitChi2', 'psfFitRedChi2'])
+
+    def __init__(self, diffim, posImage=None, negImage=None):
+        """Algorithm to run dipole measurement on a diaSource
+        @param diffim Exposure on which the diaSources were detected
+        @param posImage "Positive" exposure from which the template was subtracted
+        @param negImage "Negative" exposure which was subtracted from the posImage
+        """
+
+        self.diffim = diffim
+        self.posImage = posImage
+        self.negImage = negImage
+        self.psfSigma = None
+        if diffim is not None:
+            self.psfSigma = diffim.getPsf().computeShape().getDeterminantRadius()
+        self.log = pexLog.Log(pexLog.Log.getDefaultLog(),
+                              'lsst.ip.diffim.DipoleFitAlgorithm', pexLog.Log.INFO)
+
+        import lsstDebug
+        self.debug = lsstDebug.Info(__name__).debug
+
+    def genBgGradientModel(self, in_x, pars=()):
+        """Generate gradient model (2-d array) with up to 2nd-order polynomial
+        @param in_x Provides the input x,y grid upon which to compute the gradient
+        @param pars Up to 6 floats for up to 6 2nd-order 2-d polynomial gradient parameters.
+        If emtpy, do nothing (for speed).
+
+        @return None, or 2-d numpy.array of width/height matching
+        input bbox, containing computed gradient values.
```

Thank you for documenting the return values. But...why is silent failure supported? You don't even log it. I fear this will make it difficult to diagnose problems, since the user will have no idea if a background was subtracted or not. I recommend that you log a warning if really want to plow ahead without a gradient, and otherwise raise an exception.